File: Catalog core/src/main/java/cz/vhromada/catalog/facade/impl/ShowFacadeImpl.java
package cz.vhromada.catalog.facade.impl;

import java.util.List;

import cz.vhromada.catalog.commons.Time;
import cz.vhromada.catalog.entities.Episode;
import cz.vhromada.catalog.entities.Genre;
import cz.vhromada.catalog.entities.Season;
import cz.vhromada.catalog.entities.Show;
import cz.vhromada.catalog.facade.ShowFacade;
import cz.vhromada.catalog.facade.to.GenreTO;
import cz.vhromada.catalog.facade.to.ShowTO;
import cz.vhromada.catalog.facade.validators.ShowTOValidator;
import cz.vhromada.catalog.service.CatalogService;
import cz.vhromada.converters.Converter;
import cz.vhromada.validators.Validators;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Component;

/**
 * A class represents implementation of facade for shows.
 *
 * @author Vladimir Hromada
 */
@Component("showFacade")
public class ShowFacadeImpl implements ShowFacade {

    /**
     * TO for show argument
     */
    private static final String SHOW_TO_ARGUMENT = "TO for show";

    /**
     * TO for genre argument
     */
    private static final String GENRE_TO_ARGUMENT = "TO for genre";

    /**
     * Service for shows
     */
    private CatalogService<Show> showService;

    /**
     * Service for genres
     */
    private CatalogService<Genre> genreService;

    /**
     * Converter
     */
    private Converter converter;

    /**
     * Validator for TO for show
     */
    private ShowTOValidator showTOValidator;

    /**
     * Creates a new instance of ShowFacadeImpl.
     *
     * @param showService     service for shows
     * @param genreService    service for genres
     * @param converter       converter
     * @param showTOValidator validator for TO for show
     * @throws IllegalArgumentException if service for shows is null
     *                                  or service for genres is null
     *                                  or converter is null
     *                                  or validator for TO for show is null
     */
    @Autowired
    public ShowFacadeImpl(final CatalogService<Show> showService,
            final CatalogService<Genre> genreService,
            @Qualifier("catalogDozerConverter") final Converter converter,
            final ShowTOValidator showTOValidator) {
        Validators.validateArgumentNotNull(showService, "Service for shows");
        Validators.validateArgumentNotNull(genreService, "Service for genres");
        Validators.validateArgumentNotNull(converter, "Converter");
        Validators.validateArgumentNotNull(showTOValidator, "Validator for TO for show");

        this.showService = showService;
        this.genreService = genreService;
        this.converter = converter;
        this.showTOValidator = showTOValidator;
    }

    @Override
    public void newData() {
        showService.newData();
    }

    @Override
    public List<ShowTO> getShows() {
        return converter.convertCollection(showService.getAll(), ShowTO.class);
    }

    /**
     * @throws IllegalArgumentException {@inheritDoc}
     */
    @Override
    public ShowTO getShow(final Integer id) {
        Validators.validateArgumentNotNull(id, "ID");

        return converter.convert(showService.get(id), ShowTO.class);
    }

    /**
     * @throws IllegalArgumentException                                  {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.ValidationException     {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException {@inheritDoc}
     */
    @Override
    public void add(final ShowTO show) {
        showTOValidator.validateNewShowTO(show);
        for (final GenreTO genre : show.getGenres()) {
            Validators.validateExists(genreService.get(genre.getId()), GENRE_TO_ARGUMENT);
        }

        showService.add(converter.convert(show, Show.class));
    }

    /**
     * @throws IllegalArgumentException                                  {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.ValidationException     {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException {@inheritDoc}
     */
    @Override
    public void update(final ShowTO show) {
        showTOValidator.validateExistingShowTO(show);
        Validators.validateExists(showService.get(show.getId()), SHOW_TO_ARGUMENT);
        for (final GenreTO genre : show.getGenres()) {
            Validators.validateExists(genreService.get(genre.getId()), GENRE_TO_ARGUMENT);
        }

        showService.update(converter.convert(show, Show.class));
    }

    /**
     * @throws IllegalArgumentException                                  {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.ValidationException     {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException {@inheritDoc}
     */
    @Override
    public void remove(final ShowTO show) {
        showTOValidator.validateShowTOWithId(show);
        final Show showEntity = showService.get(show.getId());
        Validators.validateExists(showEntity, SHOW_TO_ARGUMENT);

        showService.remove(showEntity);
    }

    /**
     * @throws IllegalArgumentException                              {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.ValidationException {@inheritDoc}
     */
    @Override
    public void duplicate(final ShowTO show) {
        showTOValidator.validateShowTOWithId(show);
        final Show showEntity = showService.get(show.getId());
        Validators.validateExists(showEntity, SHOW_TO_ARGUMENT);

        showService.duplicate(showEntity);
    }

    /**
     * @throws IllegalArgumentException                                  {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.ValidationException     {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException {@inheritDoc}
     */
    @Override
    public void moveUp(final ShowTO show) {
        showTOValidator.validateShowTOWithId(show);
        final Show showEntity = showService.get(show.getId());
        Validators.validateExists(showEntity, SHOW_TO_ARGUMENT);
        final List<Show> shows = showService.getAll();
        Validators.validateMoveUp(shows, showEntity, SHOW_TO_ARGUMENT);

        showService.moveUp(showEntity);
    }

    /**
     * @throws IllegalArgumentException                                  {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.ValidationException     {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException {@inheritDoc}
     */
    @Override
    public void moveDown(final ShowTO show) {
        showTOValidator.validateShowTOWithId(show);
        final Show showEntity = showService.get(show.getId());
        Validators.validateExists(showEntity, SHOW_TO_ARGUMENT);
        final List<Show> shows = showService.getAll();
        Validators.validateMoveDown(shows, showEntity, SHOW_TO_ARGUMENT);

        showService.moveDown(showEntity);
    }

    @Override
    public void updatePositions() {
        showService.updatePositions();
    }

    @Override
    public Time getTotalLength() {
        int totalLength = 0;
        for (final Show show : showService.getAll()) {
            for (final Season season : show.getSeasons()) {
                for (final Episode episode : season.getEpisodes()) {
                    totalLength += episode.getLength();
                }
            }
        }

        return new Time(totalLength);
    }

    @Override
    public int getSeasonsCount() {
        int seasonsCount = 0;
        for (final Show show : showService.getAll()) {
            seasonsCount += show.getSeasons().size();
        }

        return seasonsCount;
    }

    @Override
    public int getEpisodesCount() {
        int episodesCount = 0;
        for (final Show show : showService.getAll()) {
            for (final Season season : show.getSeasons()) {
                episodesCount += season.getEpisodes().size();
            }
        }

        return episodesCount;
    }

}

File: catalog-core/src/main/kotlin/com/github/vhromada/catalog/facade/impl/ShowFacadeImpl.kt
package com.github.vhromada.catalog.facade.impl

import com.github.vhromada.catalog.entity.Genre
import com.github.vhromada.catalog.entity.Picture
import com.github.vhromada.catalog.entity.Show
import com.github.vhromada.catalog.facade.ShowFacade
import com.github.vhromada.common.entity.Time
import com.github.vhromada.common.facade.AbstractParentFacade
import com.github.vhromada.common.mapper.Mapper
import com.github.vhromada.common.result.Result
import com.github.vhromada.common.service.ParentService
import com.github.vhromada.common.validator.Validator
import org.springframework.stereotype.Component

/**
 * A class represents implementation of facade for shows.
 *
 * @author Vladimir Hromada
 */
@Component("showFacade")
class ShowFacadeImpl(
    private val showService: ParentService<com.github.vhromada.catalog.domain.Show>,
    private val pictureService: ParentService<com.github.vhromada.catalog.domain.Picture>,
    private val genreService: ParentService<com.github.vhromada.catalog.domain.Genre>,
    mapper: Mapper<Show, com.github.vhromada.catalog.domain.Show>,
    showValidator: Validator<Show, com.github.vhromada.catalog.domain.Show>,
    private val pictureValidator: Validator<Picture, com.github.vhromada.catalog.domain.Picture>,
    private val genreValidator: Validator<Genre, com.github.vhromada.catalog.domain.Genre>
) : AbstractParentFacade<Show, com.github.vhromada.catalog.domain.Show>(parentService = showService, mapper = mapper, validator = showValidator), ShowFacade {

    override fun updateData(data: Show): Result<Unit> {
        val storedShow = showService.get(id = data.id!!)
        val validationResult = validator.validateExists(data = storedShow)
        val pictureValidationResult = getPicture(show = data)
        val genresValidationResult = getGenres(show = data)
        if (validationResult.isOk() && pictureValidationResult.isOk() && genresValidationResult.isOk()) {
            val show = mapper.map(source = data)
                .copy(seasons = storedShow.get().seasons)
            show.createdUser = storedShow.get().createdUser
            show.createdTime = storedShow.get().createdTime
            showService.update(data = show)
        }
        return Result.of(validationResult, pictureValidationResult, genresValidationResult)
    }

    @Suppress("DuplicatedCode")
    override fun addData(data: Show): Result<Unit> {
        val pictureValidationResult = getPicture(show = data)
        val genresValidationResult = getGenres(show = data)
        if (pictureValidationResult.isOk() && genresValidationResult.isOk()) {
            val show = mapper.map(source = data)
                .copy(genres = genresValidationResult.data!!)
            showService.add(data = show)
        }
        return Result.of(pictureValidationResult, genresValidationResult)
    }

    override fun getTotalLength(): Result<Time> {
        return Result.of(data = Time(length = showService.getAll().sumBy { it.seasons.sumBy { s -> s.episodes.sumBy { e -> e.length } } }))
    }

    override fun getSeasonsCount(): Result<Int> {
        return Result.of(data = showService.getAll().sumBy { it.seasons.size })
    }

    override fun getEpisodesCount(): Result<Int> {
        return Result.of(data = showService.getAll().sumBy { it.seasons.sumBy { s -> s.episodes.size } })
    }

    /**
     * Returns result with picture and validation errors.
     *
     * @param show show
     * @returns result with picture and validation errors
     */
    @Suppress("DuplicatedCode")
    private fun getPicture(show: Show): Result<com.github.vhromada.catalog.domain.Picture> {
        if (show.picture == null) {
            return Result()
        }
        val picture = pictureService.get(id = show.picture)
        val validationResult = pictureValidator.validateExists(data = picture)
        return if (validationResult.isOk()) {
            Result.of(data = picture.get())
        } else {
            Result.of(validationResult)
        }
    }

    /**
     * Returns result with genres and validation errors.
     *
     * @param show show
     * @returns result with genres and validation errors
     */
    @Suppress("DuplicatedCode")
    private fun getGenres(show: Show): Result<MutableList<com.github.vhromada.catalog.domain.Genre>> {
        val genres = mutableListOf<com.github.vhromada.catalog.domain.Genre>()
        val result = Result.of(data = genres)
        show.genres!!.filterNotNull()
            .map { genreService.get(id = it.id!!) }
            .forEach {
                val validationResult = genreValidator.validateExists(data = it)
                if (validationResult.isOk()) {
                    genres.add(it.get())
                } else {
                    result.addEvents(eventList = validationResult.events())
                }
            }
        return result
    }

}
